source/common/stats/lazy_init.h

```diff
@@ -0,0 +1,35 @@
+#pragma once
+
+#include "source/common/common/thread.h"
+
+namespace Envoy {
+namespace Stats {
+
+/**
+ * Lazy-initialization wrapper for StatsStructType, intended for deferred instantiation of a block
+ * of stats that might not be needed in a given Envoy process.
+ *
+ * This class is thread-safe -- instantiations can occur on multiple concurrent threads.
+ */
+template <typename StatsStructType> class LazyInit {
+public:
+  // Capture the stat names object and the scope with a ctor, that can be used to instantiate a
+  // StatsStructType object later.
+  // Caller should make sure scope and stat_names outlive this object.
+  LazyInit(Stats::Scope& scope, const typename StatsStructType::StatNameType& stat_names)
+      : ctor_([&scope, &stat_names]() -> StatsStructType* {
+          return new StatsStructType(stat_names, scope);
```

Comment:
WDYT of capturing `this` and assigning ctor_ to nullptr. No need to save the lambda and its context once the stats are instantiated. For paranoia I might save the StatsStructType* in a temp before nulling ctor_.

Reply:
not sure how much that would save, but sure.

Reply:
Yeah my thinking now is that most people will *not* want this until we understand the dynamics better. nulling out the ctor reduces the overhead everyone incurs from making laziness possible.

Reply:
Done. 

as communicated offline, the work has to be done in AtomicPtr or AtomicArray.